stackslib/src/net/inv/nakamoto.rs

```diff
@@ -233,3 +242,569 @@ impl InvGenerator {
+    /// Time of last update, in seconds
+    pub last_updated_at: u64,
+    /// Burn block height of first sortition
+    pub first_block_height: u64,
+    /// Length of reward cycle
+    pub reward_cycle_len: u64,
+    /// Which neighbor is this for
+    pub neighbor_address: NeighborAddress,
+
+    /// The fields below are used for synchronizing this particular peer's inventories.
+    /// Currently tracked reward cycle
+    pub cur_reward_cycle: u64,
+    /// Status of this node.
+    /// True if we should keep talking to it; false if not
+    pub online: bool,
+    /// Last time we began talking to this peer
+    pub start_sync_time: u64,
+}
+
+impl NakamotoTenureInv {
+    pub fn new(
+        first_block_height: u64,
+        reward_cycle_len: u64,
+        neighbor_address: NeighborAddress,
+    ) -> Self {
+        Self {
+            state: NakamotoInvState::GetNakamotoInvBegin,
+            tenures_inv: BTreeMap::new(),
+            highest_sortition: 0,
+            last_updated_at: 0,
+            first_block_height,
+            reward_cycle_len,
+            neighbor_address,
+            cur_reward_cycle: 0,
+            online: true,
+            start_sync_time: 0,
+        }
+    }
+
+    /// Does this remote neighbor have the ith tenure data for the given (absolute) burn block height?
+    /// (note that block_height is the _absolute_ block height)
+    pub fn has_ith_tenure(&self, burn_block_height: u64) -> bool {
+        if burn_block_height < self.first_block_height {
+            return false;
+        }
+
+        let Some(reward_cycle) = PoxConstants::static_block_height_to_reward_cycle(
+            burn_block_height,
+            self.first_block_height,
+            self.reward_cycle_len,
+        ) else {
+            return false;
+        };
+
+        let Some(rc_tenures) = self.tenures_inv.get(&reward_cycle) else {
+            return false;
+        };
+
+        let sortition_height = burn_block_height - self.first_block_height;
+        let rc_height = sortition_height % self.reward_cycle_len;
+
+        let idx =
+            usize::try_from(rc_height / 8).expect("FATAL: reward cycle length exceeds host usize");
+        let bit = rc_height % 8;
+
+        rc_tenures
+            .get(idx)
+            .map(|bits| bits & (1 << bit) != 0)
+            .unwrap_or(false)
+    }
+
+    /// How many reward cycles of data do we have for this peer?
+    pub fn highest_reward_cycle(&self) -> u64 {
+        let Some((highest_rc, _)) = self.tenures_inv.last_key_value() else {
+            return 0;
+        };
+        *highest_rc
+    }
+
+    /// Add in a newly-discovered inventory.
+    /// NOTE: inventories are supposed to be aligned to the reward cycle
+    /// Returns true if we learned about at least one new tenure-start block
+    /// Returns false if not.
+    pub fn merge_tenure_inv(
+        &mut self,
+        tenure_inv: Vec<u8>,
+        tenure_bitlen: u16,
+        reward_cycle: u64,
+    ) -> bool {
```

Comment:
1. Yes, this is buggy; will fix.  `self.highest_sortition` is the number of sortitions represented by this state machine.  It should be equal to (reward_cycle_len * highest_queried_rc + bitlen_of_highest_queried_rc)

2. The bitvectors themselves are not merged because `tenure_inv` represents the _latest_ knowledge of blocks for the given reward cycle.  If a bit in `tenure_inv` has flipped from 1 to 0 relative to `cur_inv`, or if `tenure_inv` is shorter than `cur_inv`, then that's still pertinent information, since the block downloader ultimately consumes this information to determine which tenures to fetch from this peer.  If a tenure somehow became unavailable on the remote peer, then the block downloader will need to know that so it won't go and request it.